oeps/oep-0028-content-theming-in-react.rst

```diff
@@ -0,0 +1,206 @@
+| Title           | Content Theming in React                                       |
++-----------------+----------------------------------------------------------------+
+| Last Modified   | 2018-08-17                                                     |
++-----------------+----------------------------------------------------------------+
+| Authors         | Braden MacDonald                                               |
++-----------------+----------------------------------------------------------------+
+| Arbiter         |                                                                |
++-----------------+----------------------------------------------------------------+
+| Status          | Draft                                                          |
++-----------------+----------------------------------------------------------------+
+| Type            | Architecture                                                   |
++-----------------+----------------------------------------------------------------+
+| Created         | 2018-08-17                                                     |
++-----------------+----------------------------------------------------------------+
+| `Review Period` | 2018-08-17 - 2018-08-25                                        |
++-----------------+----------------------------------------------------------------+
+
+Context
+-------
+
+This section describes the forces at play, including technological, political,
+social, and project local. These forces are probably in tension, and should
+be called out as such. The language in this section is value-neutral. It is
+simply describing facts.
+
+Decision
+--------
+
+This section describes our response to the forces described in the Context.
+It is stated in full sentences, with active voice. "We will ..."
+
+Broad Points
+-------------
+
+NOTE: This is just a heading to list down my findings/observations for the proposal.
+Once final, this will be removed
+
+From Braden's `proposal<https://gist.github.com/bradenmacdonald/34bf8bfc8f737f9b23a2664091b8f115>_`,
+we get the following points which are further elaborated below
+
+1. Build the UI out of small, modular React components as much as possible
+
+A `God component<https://en.wikipedia.org/wiki/God_object>_` tends to know and do everything within the application. No component should be a God component. Each God component should be broken down into smaller components. These small and modular components should follow the Single responsibility principle.
+
+> A component has a single responsibility when it has one reason to change.
+
+The above definition is unambiguous because there can be lots of ways to argue whether a component has more than one responsibility or not but it becomes easier to decide when this is conveyed in terms of change. The change can be cosmetic like color, spacing, animation, etc or functional like adding new feature etc.
```

Comment:
I honestly don't think this is unambiguous. Can you give an example of such a component and the one reason it would change?

Reply:
As per my understanding a classic example of this can be `Button`.
A Button can be considered to have the following responsibilites (or change points)
* its color (styling)
* any icon (styling)
* its type or action (submit, normal) (functional)

Now if we have a single component Button, we can clearly see that it has more than resposibilities to change. Somebody can change its action, someone else can change the color and someone may add icon. Though these all is configurable through props, but you will end up writing a lot of logic in code, which will make the component difficult to update you in future. So in this case, if we were to apply Single Responsibility Principle, we would have the following components

* Button
* ColorButton
* IconButton
* ActionButton

Reply:
You've just listed multiple "responsibilities" and multiple reasons to change for a simple Button. So what if I want to change add a button that has a different colour, an Icon (animated) and has an action? Do I then end up with AnimatedIconColourActionButton?

In the above case, how would you compose an ActionButton out of a regular Button if the regular button doesn't provide ways to change it? If the regular button allows you to specify the Action, why would you use an ActionButton?

Reply:
If we follow composition, we would end up with "AnimatedIconColourActionButton". 

There can be two ways to write such a button.
* One can be to write a single component which has support to do everything based on the props(and some conditional checks).
* The other can be to compose(composition here will actually require support via props and props.children).

To answer about how would you compose ActionButton, it won't be possible to if the original Button component doesn't have support.

My intent here was to highlight this that we should have components which are small and do only one thing. But I ended up giving a bad example. Sorry for the confusion.

But this should be kept in mind, that there will be times, where we will be in dilemma of either keeping a component as a single big component (with conditionals checks ) or creating a big component via composition of smaller components.